llvm/lib/Transforms/Scalar/FPBuiltinFnSelection.cpp

```diff
@@ -106,6 +107,51 @@ static bool replaceWithLLVMIR(FPBuiltinIntrinsic &BuiltinCall) {
   return true;
 }
 
+// This function lowers llvm.fpbuiltin. intrinsic functions with max-error
+// attribute to the appropriate nvvm approximate intrinsics if it's possible.
+// If it's not possible - fallback to instruction or standard C/C++ library LLVM
+// intrinsic.
+static bool
+replaceWithApproxNVPTXCallsOrFallback(FPBuiltinIntrinsic &BuiltinCall,
+                                      std::optional<float> Accuracy) {
+  IRBuilder<> IRBuilder(&BuiltinCall);
+  SmallVector<Value *> Args(BuiltinCall.args());
+  Value *Replacement = nullptr;
+  auto *Type = BuiltinCall.getType();
+  // For now only add lowering for fdiv and sqrt. Yet nvvm intrinsics have
+  // approximate variants for sin, cos, exp2 and log2.
+  // For vector fpbuiltins for NVPTX target we don't have nvvm intrinsics,
+  // fallback to instruction or standard C/C++ library LLVM intrinsic. Also
+  // nvvm fdiv and sqrt intrisics support only float type, so fallback in this
+  // case as well.
+  switch (BuiltinCall.getIntrinsicID()) {
+  case Intrinsic::fpbuiltin_fdiv:
+    if (Accuracy.value() != 2.5)
```

Comment:
There was actually a bug - for fdiv I should have checked if max-error is 2.5, not 3.0

Reply:
The accuracy check here should be a >= check, not a != check.

Reply:
To double check, that we are on the same page, in case if max-error >= 2.5 like this it will fall back to alt math impl. I guess we would want an opposite behavior, when if max-error >= 2.5 -> replace with approx nvvm intrinsic, right?

Reply:
Actually, you are right, with an extension, that it should be just `>` check (I still need to catch `2.5` and `3.0` values to resolve issue in https://github.com/intel/llvm/pull/15836 )

Reply:
Actually, let me rewrite the patch.

Reply:
Align with nvptx spec. Also added lowering for precise fdiv/sqrt